### ISSUE TITLE
Exception when a supplied 'ping' function takes a long time

### ISSUE THREAD

**Issue by asafyish** [myndzi/pool2#10]
Hi,

I noticed that if pool2 is giving a ping function, instead of the default one, I get an exception.
/node_modules/knex/node_modules/pool2/lib/pool.js:325
ug('Allocating resource to request; waited %ds', ((new Date()) - req.ts) / 100
                                                                    ^
TypeError: Cannot read property 'ts' of undefined
    at Pool.<anonymous> (/node_modules/knex/node_modules/pool2/lib/pool.js:325:88)
    at process._tickDomainCallback (node.js:381:11)

The ping function I am using:

``` js
ping: function (res, callback) {
   // Async function. lets say it takes about 5 seconds to complete
   res.ping(callback);
}
```

Then in pool.js line 324, req is null which causes an exception.
Even though there is a check in the beginning of the function _maybeAllocateResource

``` js
if (this.requests.length === 0) { return; }
```

That's because _ping is async, and by the time it's calling the callback, this.requests is already empty.
The fix is to add the above check inside the callback, like so:

``` js
if (this.requests.length === 0) { return; }
var req = this.requests.shift();
debug('Allocating resource to request; waited %ds', ((new Date()) - req.ts) / 1000);
```

Attached a PR to fix it.

**Comment by asafyish** [comment]
Ignore #b3b9cea.

**Comment by myndzi** [comment]
This is a bit strange, since every resource allocation should be "gated" by a ping. There's no reason that requests should "lose" an item while waiting for the ping to resolve. Suspect a deeper issue, will investigate.

**Comment by myndzi** [comment]
My guess is that perhaps `_maybeAllocateResource` is getting called more times than the number of available requests; this is conceivable because of the separation between the request queue and allocations; for example, `_maybeAllocateResource` gets called on the sync interval.

I think the only clean way to solve it is to "check out" the request that the ping is intended to serve, and if the ping fails to unshift it back onto the stack.

Any chance you could provide either a snippit that can reproduce or logs from DEBUG=pool2 so that I can try to determine what specifically is going on?

**Comment by asafyish** [comment]
Hi,

Just tried the new version, 1.1.2 and it's still happening :(
Here is a code sample:

``` js
'use strict';

let knex = require('knex')({
    client: 'mysql',
    connection: {
        host: 'your-host',
        port: '3306',
        user: 'user',
        password: 'pass',
        database: 'db_name'
    },

   // When this is commented out, the select works, but idle connections drop
    pool: {
        ping: function (res, callback) {
                        // It can take a few milliseconds to finish
            res.ping(callback);
        }
    }
});

console.log('Fetching data')
for (let i = 0; i < 3; i++) {
  knex.select('name').from('campaigns').where({
    idsite: i
  }).then(function(rows) {
    console.log(rows);
  });
}
```

A few things to notice:
1. I am not using pool directly, but through knex.
2. If I remove the ping function, then the code works correctly, but long idle connections are later on not available.
3. The default "ping" implementation calls the callback immediately.

**Comment by myndzi** [comment]
Nothing changed in the new version that would fix the problem; I only added resource tracking in debug mode to help determine what's actually going with resources for a different issue.

What specifically do you mean by "idle connections drop"?

Is your mysql instance local or remote?

**Comment by myndzi** [comment]
If you can reproduce the problem while running with the environment variable `DEBUG=pool2` and post the log messages here, it would be extremely helpful.

**Comment by asafyish** [comment]
Fetching data
  pool2 Attempting to acquire minimum resources (cur=0, min=2) +0ms
  pool2 Attempting to acquire resource (cur=0, ac=0) +2ms
  pool2 Attempting to acquire resource (cur=0, ac=1) +66ms
  pool2 Growing pool: no resource to serve request +2ms
  pool2 Attempting to acquire resource (cur=0, ac=2) +0ms
  pool2 Growing pool: no resource to serve request +0ms
  pool2 Attempting to acquire resource (cur=0, ac=3) +0ms
  pool2 Growing pool: no resource to serve request +1ms
  pool2 Attempting to acquire resource (cur=0, ac=4) +0ms
  pool2 Successfully allocated new resource (cur=0, ac=4, id=0) +1s
  pool2 Successfully allocated new resource (cur=1, ac=3, id=1) +2ms
  pool2 Successfully allocated new resource (cur=2, ac=2, id=2) +1ms
  pool2 Successfully allocated new resource (cur=3, ac=1, id=3) +2ms
  pool2 Successfully allocated new resource (cur=4, ac=0, id=4) +4ms
  pool2 Allocating resource (id=0) to request; waited 1.385s +145ms
  pool2 Allocating resource (id=1) to request; waited 1.388s +5ms
  pool2 Allocating resource (id=2) to request; waited 1.39s +3ms
/Users/asaf/Dev/test-pool/node_modules/mysql/lib/protocol/Parser.js:82
        throw err;
              ^
TypeError: Cannot read property 'ts' of undefined
    at Pool.<anonymous> (/Users/asaf/Dev/test-pool/node_modules/knex/node_modules/pool2/lib/pool.js:351:108)
    at Ping.Sequence.end (/Users/asaf/Dev/test-pool/node_modules/mysql/lib/protocol/sequences/Sequence.js:96:24)
    at Ping.Sequence.OkPacket (/Users/asaf/Dev/test-pool/node_modules/mysql/lib/protocol/sequences/Sequence.js:105:8)
    at Protocol._parsePacket (/Users/asaf/Dev/test-pool/node_modules/mysql/lib/protocol/Protocol.js:271:23)
    at Parser.write (/Users/asaf/Dev/test-pool/node_modules/mysql/lib/protocol/Parser.js:77:12)
    at Protocol.write (/Users/asaf/Dev/test-pool/node_modules/mysql/lib/protocol/Protocol.js:39:16)
    at Socket.<anonymous> (/Users/asaf/Dev/test-pool/node_modules/mysql/lib/Connection.js:96:28)
    at Socket.emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at TCP.onread (net.js:538:20)

**Comment by myndzi** [comment]
Okay, great! That exactly answers my question.

There seem to be two problems here:

One, pool-minimum-resource allocations that are in progress should be counted towards the number of "will be available" resources to serve requests. This would result in only three allocations being made instead of five, since you only actually made three requests.

Two, for some reason "this._maybeAllocateResource()" is getting called on all five of those -- or at least four -- and it shouldn't be. (https://github.com/myndzi/pool2/blob/master/lib/pool.js#L444-L451)

I'll address this ASAP.

**Comment by myndzi** [comment]
Published 1.1.3, please verify that it fixes your problem :)

**Comment by asafyish** [comment]
Verified. Works great.
